src/units/abilities.cpp

```diff
@@ -885,33 +879,118 @@ std::string attack_type::weapon_specials(bool only_active, bool is_backstab) con
+	std::set<std::string> checking_name;
+	for (const config::any_child sp : specials_.all_children_range()) {
+		if((checking_tags.count(sp.key) != 0)){
+			const bool active = special_active(sp.cfg, AFFECT_SELF, sp.key);
+			add_name(weapon_abilities, active, sp.cfg["name"].str(), checking_name);
 		}
-		const auto adjacent = get_adjacent_tiles(self_loc_);
+	}
+	if(self_){
+		weapon_specials_impl_self(weapon_abilities, self_, shared_from_this(), other_attack_, self_loc_, AFFECT_SELF, checking_name, checking_tags, true);
+		if(!weapon_abilities.empty()){
+			self_abilities += translation::dsgettext("wesnoth", "Self: ");
+			self_abilities += weapon_abilities;
+			add_string(weapon_abilities, self_abilities, self_teacher_allies, checking_name);
+		}
+		weapon_specials_impl_adj(weapon_abilities, self_, shared_from_this(), other_attack_, self_loc_, AFFECT_SELF, checking_name, checking_tags, "affect_allies", true);
+		if(!weapon_abilities.empty()){
+			self_teacher_allies += translation::dsgettext("wesnoth", "Teachers: ");
+			self_teacher_allies += weapon_abilities;
+			add_string(weapon_abilities, self_abilities, self_teacher_allies, checking_name);
+		}
+		weapon_specials_impl_adj(weapon_abilities, self_, shared_from_this(), other_attack_, self_loc_, AFFECT_SELF, checking_name, checking_tags, "affect_enemies", true);
+		if(!weapon_abilities.empty()){
+			self_teacher_enemies += translation::dsgettext("wesnoth", "Enemies Teachers: ");
+			self_teacher_enemies += weapon_abilities;
+			add_string(weapon_abilities, self_abilities, self_teacher_enemies, checking_name);
+		}
+	}
+
+	if(other_) {
+		if(other_attack_) {
+			for (const config::any_child sp : other_attack_->specials_.all_children_range()) {
+				if((checking_tags.count(sp.key) != 0)){
+					const bool active = other_attack_->special_active(sp.cfg, AFFECT_OTHER, sp.key);
+					add_name(weapon_abilities, active, sp.cfg["name"].str(), checking_name);
+				}
+			}
+		}
+		weapon_specials_impl_self(weapon_abilities, other_, other_attack_, shared_from_this(), other_loc_, AFFECT_OTHER, checking_name, checking_tags);
+		weapon_specials_impl_adj(weapon_abilities, other_, other_attack_, shared_from_this(), other_loc_, AFFECT_OTHER, checking_name, checking_tags);
+		if(!weapon_abilities.empty()){
+			opponent_abilities += translation::dsgettext("wesnoth", "Opponent: ");
+			opponent_abilities += weapon_abilities;
+			add_string(weapon_abilities, self_abilities, opponent_abilities, checking_name);
+		}
+	}
+	weapon_abilities = self_abilities;
+	return weapon_abilities;
+}
+
+void attack_type::weapon_specials_impl_self(std::string& weapon_abilities, unit_const_ptr self, const_attack_ptr self_attack, const_attack_ptr other_attack, const map_location& self_loc, AFFECTS whom,
+                                     std::set<std::string>& checking_name, const std::set<std::string> checking_tags, bool leader_bool)
+{
+	if(self){
```

Comment:
excuse me for this condition, this is because i forget what this in attack dialog, self_ and other_ are ever true, but i had implement that by reflexe.

Reply:
but i will let this condition, else i fear what building not pass checking